File: sdks/ios/Sources/MRRCat/MRRCat.swift
import Foundation
import StoreKit

/// MRRCat SDK for iOS
/// Unified subscription management across platforms
@available(iOS 14.0, macOS 11.0, tvOS 14.0, watchOS 7.0, *)
public final class MRRCat: @unchecked Sendable {

    // MARK: - Singleton

    /// Shared instance of MRRCat
    public static let shared = MRRCat()

    // MARK: - Properties

    private var apiKey: String?
    private var appUserID: String?
    private var baseURL: String = ""
    private var isConfigured: Bool = false
    private var cachedSubscriberInfo: SubscriberInfo?
    private var cacheExpiry: Date?
    private let cacheDuration: TimeInterval = 300 // 5 minutes

    /// Delegate for receiving updates
    public weak var delegate: MRRCatDelegate?

    /// Current subscriber info (cached)
    public var subscriberInfo: SubscriberInfo? {
        return cachedSubscriberInfo
    }

    /// Current app user ID
    public var currentAppUserID: String? {
        return appUserID
    }

    // MARK: - Configuration

    private init() {}

    /// Configure MRRCat with your API key
    /// - Parameters:
    ///   - apiKey: Your MRRCat API key
    ///   - baseURL: Your MRRCat API URL (e.g., "https://mrrcat.yourdomain.com")
    ///   - appUserID: Optional user ID. If nil, an anonymous ID will be generated
    public func configure(
        apiKey: String,
        baseURL: String,
        appUserID: String? = nil
    ) {
        self.apiKey = apiKey
        self.baseURL = baseURL
        self.appUserID = appUserID ?? generateAnonymousID()
        self.isConfigured = true

        // Start listening for StoreKit transactions
        startTransactionListener()

        // Fetch initial subscriber info
        Task {
            try? await refreshSubscriberInfo()
        }
    }

    /// Change the current app user ID (for login/logout)
    /// - Parameter appUserID: New user ID, or nil for anonymous
    public func identify(appUserID: String?) async throws {
        self.appUserID = appUserID ?? generateAnonymousID()
        self.cachedSubscriberInfo = nil
        self.cacheExpiry = nil
        try await refreshSubscriberInfo()
    }

    /// Reset to anonymous user (logout)
    public func logOut() async throws {
        try await identify(appUserID: nil)
    }

    // MARK: - Subscriber Info

    /// Get current subscriber info
    /// - Parameter forceRefresh: Force refresh from server
    /// - Returns: SubscriberInfo
    public func getSubscriberInfo(forceRefresh: Bool = false) async throws -> SubscriberInfo {
        guard isConfigured, let _ = apiKey, let userID = appUserID else {
            throw MRRCatError.notConfigured
        }

        // Return cached if valid
        if !forceRefresh,
           let cached = cachedSubscriberInfo,
           let expiry = cacheExpiry,
           Date() < expiry {
            return cached
        }

        return try await refreshSubscriberInfo()
    }

    /// Check if user has active entitlement
    /// - Parameter identifier: Entitlement identifier (e.g., "premium")
    /// - Returns: True if entitlement is active
    public func hasEntitlement(_ identifier: String) async throws -> Bool {
        let info = try await getSubscriberInfo()
        return info.entitlements[identifier]?.isActive ?? false
    }

    @discardableResult
    private func refreshSubscriberInfo() async throws -> SubscriberInfo {
        guard let userID = appUserID else {
            throw MRRCatError.notConfigured
        }

        let info = try await apiRequest(
            method: "GET",
            path: "/v1/subscribers/\(userID)"
        ) as SubscriberResponse

        let subscriberInfo = SubscriberInfo(from: info.subscriber)
        self.cachedSubscriberInfo = subscriberInfo
        self.cacheExpiry = Date().addingTimeInterval(cacheDuration)

        DispatchQueue.main.async {
            self.delegate?.mrrcat(self, didReceiveUpdated: subscriberInfo)
        }

        return subscriberInfo
    }

    // MARK: - Purchases

    /// Purchase a product
    /// - Parameter productID: StoreKit product identifier
    /// - Returns: Updated SubscriberInfo after purchase
    @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
    public func purchase(productID: String) async throws -> SubscriberInfo {
        // Get product from StoreKit
        let products = try await Product.products(for: [productID])
        guard let product = products.first else {
            throw MRRCatError.productNotFound(productID)
        }

        return try await purchase(product: product)
    }

    /// Purchase a StoreKit Product
    /// - Parameter product: StoreKit Product
    /// - Returns: Updated SubscriberInfo after purchase
    @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
    public func purchase(product: Product) async throws -> SubscriberInfo {
        let result = try await product.purchase()

        switch result {
        case .success(let verification):
            switch verification {
            case .verified(let transaction):
                // Sync with MRRCat backend
                try await syncTransaction(transaction)
                await transaction.finish()
                return try await getSubscriberInfo(forceRefresh: true)

            case .unverified(_, let error):
                throw MRRCatError.verificationFailed(error.localizedDescription)
            }

        case .userCancelled:
            throw MRRCatError.purchaseCancelled

        case .pending:
            throw MRRCatError.purchasePending

        @unknown default:
            throw MRRCatError.unknown
        }
    }

    /// Restore purchases
    /// - Returns: Updated SubscriberInfo
    @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
    public func restorePurchases() async throws -> SubscriberInfo {
        // Sync all current entitlements
        for await result in Transaction.currentEntitlements {
            if case .verified(let transaction) = result {
                try await syncTransaction(transaction)
            }
        }

        return try await getSubscriberInfo(forceRefresh: true)
    }

    /// Sync a transaction with MRRCat backend
    @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
    private func syncTransaction(_ transaction: Transaction) async throws {
        guard let userID = appUserID else {
            throw MRRCatError.notConfigured
        }

        let body: [String: Any] = [
            "app_user_id": userID,
            "platform": "ios",
            "receipt_data": [
                "transaction_id": String(transaction.id),
                "original_transaction_id": String(transaction.originalID),
                "product_id": transaction.productID
            ]
        ]

        let _: ReceiptResponse = try await apiRequest(
            method: "POST",
            path: "/v1/receipts",
            body: body
        )
    }

    // MARK: - Products

    /// Get available products from StoreKit
    /// - Parameter productIDs: Product identifiers to fetch
    /// - Returns: Array of Products
    @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
    public func getProducts(productIDs: Set<String>) async throws -> [Product] {
        return try await Product.products(for: productIDs)
    }

    // MARK: - Offerings

    private var cachedOfferings: Offerings?
    private var offeringsCacheExpiry: Date?

    /// Get current offerings
    /// - Parameter forceRefresh: Force refresh from server
    /// - Returns: Offerings configuration
    public func getOfferings(forceRefresh: Bool = false) async throws -> Offerings {
        guard isConfigured, let _ = apiKey else {
            throw MRRCatError.notConfigured
        }

        // Return cached if valid
        if !forceRefresh,
           let cached = cachedOfferings,
           let expiry = offeringsCacheExpiry,
           Date() < expiry {
            return cached
        }

        return try await refreshOfferings()
    }

    /// Get current offering
    /// - Returns: Current offering or nil if none configured
    public func getCurrentOffering() async throws -> Offering? {
        let offerings = try await getOfferings()
        return offerings.current
    }

    /// Get offering by identifier
    /// - Parameter identifier: Offering identifier
    /// - Returns: Offering or nil if not found
    public func getOffering(identifier: String) async throws -> Offering? {
        let offerings = try await getOfferings()
        return offerings.all[identifier]
    }

    @discardableResult
    private func refreshOfferings() async throws -> Offerings {
        var queryParams = ""

        // Add targeting context
        if let userID = appUserID {
            queryParams += "?app_user_id=\(userID)"
        }

        let response = try await apiRequest(
            method: "GET",
            path: "/v1/offerings\(queryParams)"
        ) as OfferingsResponse

        let offerings = Offerings(from: response)
        self.cachedOfferings = offerings
        self.offeringsCacheExpiry = Date().addingTimeInterval(cacheDuration)

        return offerings
    }

    /// Get products for a specific package
    @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
    public func getProducts(for package: Package) async throws -> [Product] {
        let productIDs = Set(package.products.map { $0.storeProductId })
        return try await Product.products(for: productIDs)
    }

    // MARK: - Transaction Listener

    private var transactionListenerTask: Task<Void, Never>?

    private func startTransactionListener() {
        guard #available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *) else {
            return
        }

        transactionListenerTask = Task.detached { [weak self] in
            for await result in Transaction.updates {
                guard let self = self else { return }

                if case .verified(let transaction) = result {
                    do {
                        try await self.syncTransaction(transaction)
                        await transaction.finish()
                        let info = try await self.getSubscriberInfo(forceRefresh: true)

                        await MainActor.run {
                            self.delegate?.mrrcat(self, didReceiveUpdated: info)
                        }
                    } catch {
                        await MainActor.run {
                            self.delegate?.mrrcat(self, didReceiveError: error)
                        }
                    }
                }
            }
        }
    }

    // MARK: - Helpers

    private func generateAnonymousID() -> String {
        if let stored = UserDefaults.standard.string(forKey: "mrrcat_anonymous_id") {
            return stored
        }
        let id = "$anon_\(UUID().uuidString.lowercased())"
        UserDefaults.standard.set(id, forKey: "mrrcat_anonymous_id")
        return id
    }

    private func apiRequest<T: Decodable>(
        method: String,
        path: String,
        body: [String: Any]? = nil
    ) async throws -> T {
        guard let apiKey = apiKey else {
            throw MRRCatError.notConfigured
        }

        guard let url = URL(string: baseURL + path) else {
            throw MRRCatError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("MRRCat-iOS/1.0.0", forHTTPHeaderField: "User-Agent")

        if let body = body {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw MRRCatError.networkError
        }

        guard 200...299 ~= httpResponse.statusCode else {
            if let error = try? JSONDecoder().decode(APIError.self, from: data) {
                throw MRRCatError.apiError(error.error.code, error.error.message)
            }
            throw MRRCatError.httpError(httpResponse.statusCode)
        }

        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return try decoder.decode(T.self, from: data)
    }
}

// MARK: - Delegate Protocol

@available(iOS 14.0, macOS 11.0, tvOS 14.0, watchOS 7.0, *)
public protocol MRRCatDelegate: AnyObject {
    func mrrcat(_ mrrcat: MRRCat, didReceiveUpdated subscriberInfo: SubscriberInfo)
    func mrrcat(_ mrrcat: MRRCat, didReceiveError error: Error)
}

// MARK: - Default Delegate Implementation

@available(iOS 14.0, macOS 11.0, tvOS 14.0, watchOS 7.0, *)
public extension MRRCatDelegate {
    func mrrcat(_ mrrcat: MRRCat, didReceiveError error: Error) {}
}

File: sdks/flutter/lib/src/mrrcat.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:in_app_purchase/in_app_purchase.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'models.dart';
import 'exceptions.dart';

/// MRRCat Flutter SDK
class MRRCat {
  static MRRCat? _instance;
  static MRRCat get instance {
    if (_instance == null) {
      throw MRRCatNotConfiguredException();
    }
    return _instance!;
  }

  /// Check if MRRCat is configured
  static bool get isConfigured => _instance != null;

  final String _apiKey;
  String _appUserID;
  final String _baseURL;

  final InAppPurchase _iap = InAppPurchase.instance;
  StreamSubscription<List<PurchaseDetails>>? _purchaseSubscription;

  SubscriberInfo? _cachedSubscriberInfo;
  DateTime? _cacheExpiry;
  final Duration _cacheDuration = const Duration(minutes: 5);

  final _subscriberInfoController = StreamController<SubscriberInfo>.broadcast();

  /// Stream of subscriber info updates
  Stream<SubscriberInfo> get subscriberInfoStream => _subscriberInfoController.stream;

  /// Current subscriber info (cached)
  SubscriberInfo? get subscriberInfo => _cachedSubscriberInfo;

  /// Current app user ID
  String get appUserID => _appUserID;

  MRRCat._({
    required String apiKey,
    required String appUserID,
    required String baseURL,
  })  : _apiKey = apiKey,
        _appUserID = appUserID,
        _baseURL = baseURL;

  /// Configure MRRCat SDK
  static Future<MRRCat> configure({
    required String apiKey,
    String? appUserID,
    String baseURL = 'https://mrrcat.ongoru.workers.dev',
  }) async {
    if (_instance != null) {
      return _instance!;
    }

    final userID = appUserID ?? await _getOrCreateAnonymousID();

    _instance = MRRCat._(
      apiKey: apiKey,
      appUserID: userID,
      baseURL: baseURL,
    );

    await _instance!._initialize();
    return _instance!;
  }

  Future<void> _initialize() async {
    // Setup purchase listener
    _purchaseSubscription = _iap.purchaseStream.listen(
      _onPurchaseUpdated,
      onError: (error) {
        debugPrint('MRRCat: Purchase stream error: $error');
      },
    );

    // Fetch initial subscriber info
    try {
      await getSubscriberInfo();
    } catch (e) {
      debugPrint('MRRCat: Failed to fetch initial subscriber info: $e');
    }
  }

  /// Get subscriber info
  Future<SubscriberInfo> getSubscriberInfo({bool forceRefresh = false}) async {
    if (!forceRefresh &&
        _cachedSubscriberInfo != null &&
        _cacheExpiry != null &&
        DateTime.now().isBefore(_cacheExpiry!)) {
      return _cachedSubscriberInfo!;
    }

    final response = await _apiRequest('GET', '/v1/subscribers/$_appUserID');
    final data = jsonDecode(response.body);
    final info = SubscriberInfo.fromJson(data['subscriber']);

    _cachedSubscriberInfo = info;
    _cacheExpiry = DateTime.now().add(_cacheDuration);
    _subscriberInfoController.add(info);

    return info;
  }

  /// Check if user has active entitlement
  Future<bool> hasEntitlement(String identifier) async {
    final info = await getSubscriberInfo();
    return info.entitlements[identifier]?.isActive ?? false;
  }

  /// Identify user (login)
  Future<SubscriberInfo> identify(String newAppUserID) async {
    _appUserID = newAppUserID;
    _cachedSubscriberInfo = null;
    _cacheExpiry = null;
    return getSubscriberInfo();
  }

  /// Log out and switch to anonymous user
  Future<SubscriberInfo> logOut() async {
    _appUserID = await _getOrCreateAnonymousID(forceNew: true);
    _cachedSubscriberInfo = null;
    _cacheExpiry = null;
    return getSubscriberInfo();
  }

  /// Get available products
  Future<List<ProductDetails>> getProducts(Set<String> productIDs) async {
    final available = await _iap.isAvailable();
    if (!available) {
      throw MRRCatException('store_unavailable', 'In-app purchases are not available');
    }

    final response = await _iap.queryProductDetails(productIDs);
    if (response.error != null) {
      throw MRRCatException('product_query_failed', response.error!.message);
    }

    return response.productDetails;
  }

  /// Purchase a product
  Future<SubscriberInfo> purchase(ProductDetails product) async {
    final purchaseParam = PurchaseParam(productDetails: product);

    final result = await _iap.buyNonConsumable(purchaseParam: purchaseParam);
    if (!result) {
      throw MRRCatPurchaseCancelledException();
    }

    // Wait for purchase to complete via stream
    final completer = Completer<SubscriberInfo>();

    late StreamSubscription<SubscriberInfo> subscription;
    subscription = subscriberInfoStream.listen((info) {
      if (!completer.isCompleted) {
        completer.complete(info);
        subscription.cancel();
      }
    });

    // Timeout after 60 seconds
    Future.delayed(const Duration(seconds: 60), () {
      if (!completer.isCompleted) {
        subscription.cancel();
        completer.completeError(
          MRRCatException('purchase_timeout', 'Purchase timed out'),
        );
      }
    });

    return completer.future;
  }

  /// Restore purchases
  Future<SubscriberInfo> restorePurchases() async {
    await _iap.restorePurchases();
    return getSubscriberInfo(forceRefresh: true);
  }

  // Offerings
  Offerings? _cachedOfferings;
  DateTime? _offeringsCacheExpiry;

  /// Cached offerings
  Offerings? get offerings => _cachedOfferings;

  /// Get offerings
  Future<Offerings> getOfferings({bool forceRefresh = false}) async {
    if (!forceRefresh &&
        _cachedOfferings != null &&
        _offeringsCacheExpiry != null &&
        DateTime.now().isBefore(_offeringsCacheExpiry!)) {
      return _cachedOfferings!;
    }

    final response = await _apiRequest('GET', '/v1/offerings?app_user_id=$_appUserID');
    final data = jsonDecode(response.body);
    final offerings = Offerings.fromJson(data);

    _cachedOfferings = offerings;
    _offeringsCacheExpiry = DateTime.now().add(_cacheDuration);

    return offerings;
  }

  /// Get current offering
  Future<Offering?> getCurrentOffering() async {
    final offerings = await getOfferings();
    return offerings.current;
  }

  /// Get offering by identifier
  Future<Offering?> getOffering(String identifier) async {
    final offerings = await getOfferings();
    return offerings.all[identifier];
  }

  void _onPurchaseUpdated(List<PurchaseDetails> purchases) async {
    for (final purchase in purchases) {
      if (purchase.status == PurchaseStatus.purchased ||
          purchase.status == PurchaseStatus.restored) {
        try {
          await _syncPurchase(purchase);

          if (purchase.pendingCompletePurchase) {
            await _iap.completePurchase(purchase);
          }

          await getSubscriberInfo(forceRefresh: true);
        } catch (e) {
          debugPrint('MRRCat: Failed to sync purchase: $e');
        }
      } else if (purchase.status == PurchaseStatus.error) {
        debugPrint('MRRCat: Purchase error: ${purchase.error}');
      }
    }
  }

  Future<void> _syncPurchase(PurchaseDetails purchase) async {
    final platform = Platform.isIOS ? 'ios' : 'android';

    final body = {
      'app_user_id': _appUserID,
      'platform': platform,
      'receipt_data': {
        if (Platform.isIOS) 'transaction_id': purchase.purchaseID,
        if (Platform.isAndroid) 'purchase_token': purchase.verificationData.serverVerificationData,
        'product_id': purchase.productID,
      },
    };

    await _apiRequest('POST', '/v1/receipts', body: body);
  }

  Future<http.Response> _apiRequest(
    String method,
    String path, {
    Map<String, dynamic>? body,
  }) async {
    final uri = Uri.parse('$_baseURL$path');

    final headers = {
      'Content-Type': 'application/json',
      'X-API-Key': _apiKey,
      'User-Agent': 'MRRCat-Flutter/1.0.0',
    };

    http.Response response;

    switch (method) {
      case 'GET':
        response = await http.get(uri, headers: headers);
        break;
      case 'POST':
        response = await http.post(
          uri,
          headers: headers,
          body: body != null ? jsonEncode(body) : null,
        );
        break;
      default:
        throw MRRCatException('invalid_method', 'Invalid HTTP method: $method');
    }

    if (response.statusCode >= 400) {
      try {
        final error = jsonDecode(response.body);
        throw MRRCatApiException(
          error['error']['code'] ?? 'unknown',
          error['error']['message'] ?? 'Request failed',
        );
      } catch (e) {
        if (e is MRRCatException) rethrow;
        throw MRRCatHttpException(response.statusCode);
      }
    }

    return response;
  }

  static Future<String> _getOrCreateAnonymousID({bool forceNew = false}) async {
    final prefs = await SharedPreferences.getInstance();
    const key = 'mrrcat_anonymous_id';

    if (!forceNew) {
      final stored = prefs.getString(key);
      if (stored != null) return stored;
    }

    final id = '\$anon_${DateTime.now().millisecondsSinceEpoch}_${_generateRandomString(8)}';
    await prefs.setString(key, id);
    return id;
  }

  static String _generateRandomString(int length) {
    const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
    final random = DateTime.now().millisecondsSinceEpoch;
    return String.fromCharCodes(
      Iterable.generate(
        length,
        (_) => chars.codeUnitAt((random + _) % chars.length),
      ),
    );
  }

  /// Dispose resources
  void dispose() {
    _purchaseSubscription?.cancel();
    _subscriberInfoController.close();
  }
}
